service/history/query.go

```diff
@@ -0,0 +1,120 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package history
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/shared"
+)
+
+const (
+	queryStateBuffered queryState = iota
+	queryStateCompleted
+)
+
+type (
+	queryState int
+
+	query interface {
+		getQuerySnapshot() *querySnapshot
+		getQueryTermCh() <-chan struct{}
+		completeQuery(*shared.WorkflowQueryResult) error
+	}
+
+	queryImpl struct {
+		sync.RWMutex
+
+		id          string
+		queryInput  *shared.WorkflowQuery
+		queryResult *shared.WorkflowQueryResult
+		termCh      chan struct{}
+		state       queryState
+	}
```

Comment:
```go
	queryImpl struct {
		id          	string
		termCh	chan struct{}
		
		sync.RWMutex
		queryInput	*shared.WorkflowQuery
		queryResult	*shared.WorkflowQueryResult
		state       		queryState
	}
```

Reply:
generally, things are immutable should be put at the very top
things are guarded by lock should be put under the lock

Reply:
sounds good, made the change.